machinedeployment_controller fails with null pointer exception if Spec.Strategy is not defined

**What steps did you take and what happened:**

Attempted to create a simple Azure cluster and the `capi-controller-manager` pod failed with error:

```
I0421 21:54:23.910221       1 controller.go:132] controller-runtime/controller "msg"="Starting EventSource" "controller"="cluster" "source"={"Type":{"apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha3","kind":"AzureCluster"}}
E0421 21:54:23.931064       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 397 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1644240, 0x26e9e80)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1644240, 0x26e9e80)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api/controllers/mdutil.NewMSNewReplicas(0xc0003e1500, 0xc00000e7c0, 0x1, 0x1, 0xc000596000, 0xc00000e7c0, 0x0, 0x1)
        /workspace/controllers/mdutil/util.go:519 +0x165
sigs.k8s.io/cluster-api/controllers.(*MachineDeploymentReconciler).getNewMachineSet(0xc000595e80, 0xc0003e1500, 0x2743460, 0x0, 0x0, 0x2743460, 0x0, 0x0, 0x1, 0x0, ...)
        /workspace/controllers/machinedeployment_sync.go:170 +0x7f1
sigs.k8s.io/cluster-api/controllers.(*MachineDeploymentReconciler).getAllMachineSetsAndSyncRevision(0xc000595e80, 0xc0003e1500, 0x2743460, 0x0, 0x0, 0x1, 0x1a90720, 0x0, 0x0, 0x0, ...)
        /workspace/controllers/machinedeployment_sync.go:76 +0xd7
sigs.k8s.io/cluster-api/controllers.(*MachineDeploymentReconciler).rolloutRolling(0xc000595e80, 0xc0003e1500, 0x2743460, 0x0, 0x0, 0x0, 0x0)
        /workspace/controllers/machinedeployment_rolling.go:30 +0x7c
sigs.k8s.io/cluster-api/controllers.(*MachineDeploymentReconciler).reconcile(0xc000595e80, 0x1a80200, 0xc0000ae060, 0xc00009bba0, 0xc0003e1500, 0x0, 0x0, 0xc, 0xc00009bba0)
        /workspace/controllers/machinedeployment_controller.go:203 +0x7e9
sigs.k8s.io/cluster-api/controllers.(*MachineDeploymentReconciler).Reconcile(0xc000595e80, 0xc0007b0140, 0x9, 0xc00076fc20, 0x11, 0xc00047e200, 0x0, 0x0, 0x0)
        /workspace/controllers/machinedeployment_controller.go:147 +0x4e0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0004a6420, 0x169b1c0, 0xc00055c680, 0x0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:255 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0004a6420, 0x0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:231 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0004a6420)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:210 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0007fed30)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0007fed30, 0x3b9aca00, 0x0, 0x1, 0xc000472000)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0007fed30, 0x3b9aca00, 0xc000472000)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.9/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.14/pkg/internal/controller/controller.go:192 +0x468
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x14916b5]
```

**What did you expect to happen:**

Provisioning to start without errors.


**Anything else you would like to add:**

Error is in `controllers/machinedeployment_controller.go` in the `reconcile` function:

```go
...
	// d.Spec.Strategy is not set by default, it should default to RollingUpdate and this condition should check for null
	if d.Spec.Strategy.Type == clusterv1.RollingUpdateMachineDeploymentStrategyType {
		return ctrl.Result{}, r.rolloutRolling(ctx, d, msList)
	}

	if d.Spec.Strategy.Type == clusterv1.OnDeleteMachineDeploymentStrategyType {
		return ctrl.Result{}, r.rolloutOnDelete(ctx, d, msList)
	}
...

```


**Environment:**

- Cluster-api version: 0.3.16
- Minikube/KIND version: N/A
- Kubernetes version: (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T21:10:45Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/arm64"}
```

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

--- a/controllers/machinedeployment_controller.go
+++ b/controllers/machinedeployment_controller.go
@@ -195,7 +195,14 @@ func (r *MachineDeploymentReconciler) reconcile(ctx context.Context, cluster *cl
 		return ctrl.Result{}, r.sync(ctx, d, msList)
 	}
 
+	if d.Spec.Strategy == nil {
+		return ctrl.Result{}, errors.Errorf("missing MachineDeployment strategy")
+	}
+
 	if d.Spec.Strategy.Type == clusterv1.RollingUpdateMachineDeploymentStrategyType {
+		if d.Spec.Strategy.RollingUpdate == nil {
+			return ctrl.Result{}, errors.Errorf("missing MachineDeployment settings for strategy type: %s", d.Spec.Strategy.Type)
+		}
 		return ctrl.Result{}, r.rolloutRolling(ctx, d, msList)
 	}
 

